source/common/network/buffered_io_socket_handle_impl.h

```diff
@@ -0,0 +1,163 @@
+
+  bool supportsMmsg() const override;
+  bool supportsUdpGro() const override;
+
+  Api::SysCallIntResult bind(Address::InstanceConstSharedPtr address) override;
+  Api::SysCallIntResult listen(int backlog) override;
+  IoHandlePtr accept(struct sockaddr* addr, socklen_t* addrlen) override;
+  Api::SysCallIntResult connect(Address::InstanceConstSharedPtr address) override;
+  Api::SysCallIntResult setOption(int level, int optname, const void* optval,
+                                  socklen_t optlen) override;
+  Api::SysCallIntResult getOption(int level, int optname, void* optval, socklen_t* optlen) override;
+  Api::SysCallIntResult setBlocking(bool blocking) override;
+  absl::optional<int> domain() override;
+  Address::InstanceConstSharedPtr localAddress() override;
+  Address::InstanceConstSharedPtr peerAddress() override;
+  Event::FileEventPtr createFileEvent(Event::Dispatcher& dispatcher, Event::FileReadyCb cb,
+                                      Event::FileTriggerType trigger, uint32_t events) override;
+  Api::SysCallIntResult shutdown(int how) override;
+  absl::optional<std::chrono::milliseconds> lastRoundTripTime() override { return absl::nullopt; }
+
+  Buffer::WatermarkBuffer& getBufferForTest() { return owned_buffer_; }
+
+  void scheduleNextEvent() {
+    // It's possible there is no pending file event so as no io_callback.
+    if (io_callback_) {
+      ENVOY_LOG(trace, "Schedule IO callback on {}", static_cast<void*>(this));
+      io_callback_->scheduleCallbackNextIteration();
+    }
+  }
+
+  void setWritablePeer(WritablePeer* writable_peer) {
+    // Swapping writable peer is undefined behavior.
+    ASSERT(!writable_peer_);
+    ASSERT(!write_shutdown_);
+    writable_peer_ = writable_peer;
+  }
+
+  // WritablePeer
+  void setWriteEnd() override { read_end_stream_ = true; }
+  bool isWriteEndSet() override { return read_end_stream_; }
+  void maybeSetNewData() override {
+    ENVOY_LOG(trace, "{} on socket {}", __FUNCTION__, static_cast<void*>(this));
+    scheduleNextEvent();
+  }
+  void onPeerDestroy() override {
+    writable_peer_ = nullptr;
+    write_shutdown_ = true;
+  }
+  void onPeerBufferWritable() override { scheduleNextEvent(); }
+  bool isWritable() const override { return !isOverHighWatermark(); }
+  Buffer::Instance* getWriteBuffer() override { return &owned_buffer_; }
+
+  // ReadableSource
+  bool isPeerShutDownWrite() const override { return read_end_stream_; }
+  bool isOverHighWatermark() const override { return over_high_watermark_; }
+  bool isReadable() const override { return isPeerShutDownWrite() || owned_buffer_.length() > 0; }
+
+private:
+  // Support isOpen() and close(). IoHandle owner must invoke close() to avoid potential resource
+  // leak.
+  bool closed_{false};
+
+  // The attached file event with this socket. The event is not owned by the socket.
+  Event::UserSpaceFileEventImpl* user_file_event_;
```

Comment:
Envoy prefers references where possible.

Reply:
There is discussion and experiment that if io handle could own the `event`.  But right now though event and io handle is 1:1 mapping at any given time, the event could be destroyed and reconstructed (e.g listener filter A destroy the event and listener filter B create another event after). To support this existing sequence, the `event` cannot be reference. 

I can add the comment TODO that here this io handle support and is compatible with the goal: event owned by io handle.